packages/core/src/controllers/publisher.ts

```diff
@@ -37,42 +38,57 @@ export class Publisher extends IPublisher {
   public publish: IPublisher["publish"] = async (topic, message, opts) => {
     this.logger.debug(`Publishing Payload`);
     this.logger.trace({ type: "method", method: "publish", params: { topic, message, opts } });
+
+    const ttl = opts?.ttl || PUBLISHER_DEFAULT_TTL;
+    const relay = getRelayProtocolName(opts);
+    const prompt = opts?.prompt || false;
+    const tag = opts?.tag || 0;
+    const id = opts?.id || (getBigIntRpcId().toString() as any);
+    const params = { topic, message, opts: { ttl, relay, prompt, tag, id } };
+    const failedPublishMessage = `Failed to publish payload, please try again. id:${id} tag:${tag}`;
+    const startPublish = Date.now();
+    let result;
+    let attempts = 1;
+
     try {
-      const ttl = opts?.ttl || PUBLISHER_DEFAULT_TTL;
-      const relay = getRelayProtocolName(opts);
-      const prompt = opts?.prompt || false;
-      const tag = opts?.tag || 0;
-      const id = opts?.id || (getBigIntRpcId().toString() as any);
-      const params = { topic, message, opts: { ttl, relay, prompt, tag, id } };
-      // delay adding to queue to avoid cases where heartbeat might pulse right after publish resulting in duplicate publish
-      const queueTimeout = setTimeout(() => this.queue.set(id, params), this.publishTimeout);
-      try {
+      /**
+       * Loop until the publish is successful or the timeout is reached
+       * The loop allows to retry to retry the publish in case of disconnect
+       */
+      while (result === undefined) {
+        if (Date.now() - startPublish > this.publishTimeout) {
+          throw new Error(failedPublishMessage);
+        }
+        this.logger.trace({ id, attempts }, "publisher.publish - attempt");
         const publish = await createExpiringPromise(
-          this.rpcPublish(topic, message, ttl, relay, prompt, tag, id),
+          this.rpcPublish(topic, message, ttl, relay, prompt, tag, id).catch((e) =>
+            this.logger.warn(e),
+          ),
           this.publishTimeout,
-          `Failed to publish payload, please try again. id:${id} tag:${tag}`,
+          failedPublishMessage,
         );
-        await publish;
-        this.removeRequestFromQueue(id);
-        this.relayer.events.emit(RELAYER_EVENTS.publish, params);
-      } catch (err) {
-        this.logger.debug(`Publishing Payload stalled`);
-        this.needsTransportRestart = true;
-        if (opts?.internal?.throwOnFailedPublish) {
-          // remove the request from the queue so it's not retried automatically
-          this.removeRequestFromQueue(id);
-          throw err;
+        result = await publish;
+        attempts++;
+
+        if (!result) {
+          // small buffer to allow transport to beging reconnecting
```

Comment:
Explains the "what?" but not the "why?", i.e. why is this needed?

```suggestion
          // small buffer to allow transport to begin reconnecting
```

Reply:
Also: is it sufficient under all circumstances? Can this create more edge cases if this assumption fails?

Reply:
I've updated the comment, but also its role is to avoid spam call `rpcPublish` if network is down for a few seconds but limit publish to max 1 try per second